src/algebra/module/torsion.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import algebra.module
+import linear_algebra.quotient
+import ring_theory.ideal.quotient
+import ring_theory.non_zero_divisors
+
+/-!
+# Torsion submodules
+
+## Main definitions
+
+* `torsion R M a` : the `a`-torsion submodule, containing all elements `x` of `M` such that
+  `a • x = 0`.
+* `torsion'' R M S`: the `S`-torsion submodule, containing all elements `x` of `M` such that
+  `a • x = 0` for some `a` in `S`.
+* `torsion' R M` : the torsion submoule, containing all elements `x` of `M` such that  `a • x = 0`
+  for some non-zero-divisor `a` in `R`.
+
+## Main statements
+
+* `torsion' R M` is a submodule.
+* `torsion R M a` can be viewed as a `(R ⧸ R·a)`-module.
+* `no_zero_smul_divisors_iff_torsion'_bot` : a module over a domain has `no_zero_smul_divisors`
+  (that is, there is no non-zero `a`, `x` such that `a • x = 0` )iff its `torsion'` submodule is
+  trivial.
+* `quotient_torsion.torsion'_eq_bot` : quotienting by the `tosion'` submodule makes the `torsion'`
+  submodule of the new module trivial. If `R` is a domain, we can derive an instance
+  `quotient_torsion.no_zero_smul_divisors : no_zero_smul_divisors R (M ⧸ torsion' R M)`.
+
+## Notation
+
+* The notions are defined for a `comm_semiring R` and a `module R M`. Some additional hypotheses on
+  `R` and `M` are required by some lemmas.
+* The letters `a`, `b`, ... are used for scalars (in `R`), while `x`, `y`, ... are used for vectors
+  (in `M`).
+
+## Tags
+
+Torsion, submodule, module, quotient
+-/
+
+open_locale non_zero_divisors
+section defs
+variables (R M : Type*) [comm_semiring R] [add_comm_monoid M] [module R M]
+
+/-- The `a`-torsion submodule for `a` in `R`, containing all elements `x` of `M` such that
+  `a • x = 0`. -/
+def torsion (a : R) : submodule R M := (distrib_mul_action.to_linear_map _ _ a).ker
```

Comment:
This (and other names) should go in the `submodule` namespace.

Reply:
I did it although I'm not sure what's the point of it, considering none of the functions take a `submodule R M` as argument